diff -up bind-9.7.0-P1/lib/dns/validator.c.rh554316 bind-9.7.0-P1/lib/dns/validator.c
--- bind-9.7.0-P1/lib/dns/validator.c.rh554316	2010-02-25 06:26:27.000000000 +0100
+++ bind-9.7.0-P1/lib/dns/validator.c	2010-03-26 15:37:56.113969498 +0100
@@ -2211,7 +2211,7 @@ validatezonekey(dns_validator_t *val) {
 		     result == ISC_R_SUCCESS;
 		     result = dns_rdataset_next(val->event->sigrdataset))
 		{
-			dns_keynode_t *keynode = NULL, *nextnode = NULL;
+			dns_keynode_t *keynode = NULL;
 
 			dns_rdata_reset(&sigrdata);
 			dns_rdataset_current(val->event->sigrdataset,
@@ -2231,6 +2231,7 @@ validatezonekey(dns_validator_t *val) {
 			    result == ISC_R_SUCCESS)
 				atsep = ISC_TRUE;
 			while (result == ISC_R_SUCCESS) {
+				dns_keynode_t *nextnode = NULL;
 				dstkey = dns_keynode_key(keynode);
 				if (dstkey == NULL) {
 					dns_keytable_detachkeynode(
